lib/Transforms/SimplifyParsers.cpp

```diff
@@ -0,0 +1,168 @@
+    llvm::DenseSet<P4HIR::ParserStateOp> findReachableStates(P4HIR::ParserOp &parser,
+                                                             mlir::SymbolTable &symbolTable);
+
+    /// Collapses linear sequences of states without branches or annotations.
+    void collapseChains(P4HIR::ParserOp &parser, mlir::SymbolTable &symbolTable);
+
+    /// Calculates state in-degrees and identifies unique successors.
+    void buildIndegreeAndSuccessor(
+        P4HIR::ParserOp &parser, mlir::SymbolTable &symbolTable,
+        llvm::DenseMap<P4HIR::ParserStateOp, unsigned> &indegree,
+        llvm::DenseMap<P4HIR::ParserStateOp, P4HIR::ParserStateOp> &succ);
+};
+}  // end anonymous namespace
+
+llvm::DenseSet<P4HIR::ParserStateOp> SimplifyParsers::findReachableStates(
+    P4HIR::ParserOp &parser, mlir::SymbolTable &symbolTable) {
+    auto start = symbolTable.lookup<P4HIR::ParserStateOp>("start");
+    if (!start) {
+        parser.emitOpError("Parser missing 'start' state");
+        return {};
+    }
+    llvm::SmallVector<P4HIR::ParserStateOp> worklist{start};
+    llvm::DenseSet<P4HIR::ParserStateOp> visited{start};
+
+    bool acceptReachable = false, rejectReachable = false;
+
+    while (!worklist.empty()) {
+        auto state = worklist.pop_back_val();
+        Operation *term = state.getBody().back().getTerminator();
+
+        mlir::TypeSwitch<Operation *, void>(term)
+            .Case<P4HIR::ParserAcceptOp>([&](auto) { acceptReachable = true; })
+            .Case<P4HIR::ParserRejectOp>([&](auto) { rejectReachable = true; })
+            .Case<P4HIR::ParserTransitionOp>([&](auto transition) {
+                auto next = symbolTable.lookup<P4HIR::ParserStateOp>(
+                    transition.getStateAttr().getLeafReference());
+                if (visited.insert(next).second) worklist.push_back(next);
+            })
+            .Case<P4HIR::ParserTransitionSelectOp>([&](P4HIR::ParserTransitionSelectOp select) {
+                for (auto selectCase : select.getBody().getOps<P4HIR::ParserSelectCaseOp>()) {
+                    auto next = symbolTable.lookup<P4HIR::ParserStateOp>(
+                        selectCase.getStateAttr().getLeafReference());
+                    if (visited.insert(next).second) worklist.push_back(next);
+                }
+            })
+            .Default([&](auto) { llvm_unreachable("Unknown parser terminator"); });
+    }
+
+    if (!acceptReachable && !rejectReachable)
+        parser.emitWarning("Parser never reaches the 'accept' or 'reject' state.");
+
+    return visited;
+}
+
+void SimplifyParsers::buildIndegreeAndSuccessor(
+    P4HIR::ParserOp &parser, mlir::SymbolTable &symbolTable,
+    llvm::DenseMap<P4HIR::ParserStateOp, unsigned> &indegree,
+    llvm::DenseMap<P4HIR::ParserStateOp, P4HIR::ParserStateOp> &succ) {
+    // Helper function to identify special states that should never be collapsed
+    auto isSpecial = [](P4HIR::ParserStateOp state) {
```

Comment:
name should be meaningless here (e.g. `verify` call is expanded into reject). So, we need to look into states that:
 - Has transition from the body. This will be start
 - Has accept / reject op inside.